R/ResultsDataModelQueries.R

```diff
@@ -0,0 +1,294 @@
+#'
+#' @description
+#' Given a premerged file (an .RData/rds object the output
+#' of \code{CohortDiagnostics::preMergeDiagnosticsFiles} reads the object into
+#' memory and makes it available for query.
+#'
+#' @param premergedDataFile  an .RData/rds object the output
+#'                           of \code{CohortDiagnostics::preMergeDiagnosticsFiles}
+#'
+#' @return
+#' R environment containing data conforming to Cohort Diagnostics results data model specifications.
+#' @export
+createFileDataSource <-
+  function(premergedDataFile, envir = new.env()) {
+    load(premergedDataFile, envir = envir)
+    return(envir)
+  }
+
+
+# private function - not exported
+# this function supports query with connection that is either pool or not pool. It can also establish a connection
+# when connection details is provided
+renderTranslateQuerySql <-
+  function(sql,
+           connection = NULL,
+           connectionDetails = NULL,
+           ...,
+           snakeCaseToCamelCase = FALSE) {
+    # Set up connection to server ----------------------------------------------------
+    if (is.null(connection)) {
+      if (!is.null(connectionDetails)) {
+        writeLines("Connecting to database using provided connection details.")
+        connection <- DatabaseConnector::connect(connectionDetails)
+        on.exit(DatabaseConnector::disconnect(connection))
+      } else {
+        stop("No connection or connectionDetails provided.")
+      }
+    }
+    
+    if (methods::is(connection, "Pool")) {
+      # Connection pool is used by Shiny app, which always uses PostgreSQL:
+      sql <- SqlRender::render(sql, ...)
+      sql <- SqlRender::translate(sql, targetDialect = "postgresql")
+      
+      tryCatch({
+        data <- DatabaseConnector::dbGetQuery(connection, sql)
+      }, error = function(err) {
+        writeLines(sql)
+        stop(err)
+      })
+      if (snakeCaseToCamelCase) {
+        colnames(data) <- SqlRender::snakeCaseToCamelCase(colnames(data))
+      }
+      return(data %>% dplyr::tibble())
+    } else {
+      if (is.null(attr(connection, "dbms"))) {
+        stop("dbms not provided. Unable to translate query.")
+      }
+      return(
+        DatabaseConnector::renderTranslateQuerySql(
+          connection = connection,
+          sql = sql,
+          ...,
+          snakeCaseToCamelCase = snakeCaseToCamelCase
+        ) %>% dplyr::tibble()
+      )
+    }
+  }
+
+# private function - not exported
+getDataForDatabaseIdsCohortIds <- function(dataSource,
+                                           cohortIds = NULL,
+                                           databaseIds = NULL,
+                                           dataTableName) {
+  if (any(is.null(databaseIds), length(databaseIds) == 0)) {
+    stop("Database Ids not correctly specified.")
+  }
+  
+  if (any(!exists('cohortIds'), length(cohortIds) == 0)) {
+    stop("Cohort Ids not correctly specified.")
+  }
+  
+  if (is(dataSource, "environment")) {
```

Comment:
Instead of environment a better option may be to give the dataSource an S3 class. This would be quite flexible and could be used to improve performance in the future as there maybe sections of code that would return joins from the database, rather than pulling out the table and doing it in dplyr.

For example:

```
getDataForDatabaseIdsCohortIds <- function(dataSource,...) {
    UseMethod("getDatabaseIdsCohortIds", dataSource)
}

getDatabaseCohortIds.environmentDataModel <- function(...) {
   .... # Getting data from premerged file
}

getDatabaseIdsCohortIds.PooledConnection <- function(...) {
   .... # However databases should be queried if they're pooled connections
}

x <- environment()
readRDS("PreMergedFile.rds", envir = x) 
class(x) <- "environmentDataModel"


x <- DatabaseConnector::connection(...)
```

Reply:
Interesting proposal - maybe next version? Can we make this an issue for future work?

Reply:
The main reason we went with an environment is because our Shiny apps loaded everything in the global environment, and you could just point to the global environment in these functions. Using S3 classes may be cleaner (but will require some rewrite of the Shiny app).

Reply:
Ok then definitely future version